src/vulkan/engine_vulkan_debugger.cc

```diff
@@ -0,0 +1,979 @@
+
+    dap::StackFrame frame;
+    if (!client.TopStackFrame(threadId, &frame)) {
+      return false;
+    }
+
+    dap::ScopesRequest scopeReq;
+    dap::ScopesResponse scopeRes;
+    scopeReq.frameId = frame.id;
+    if (!client.Send(scopeReq, &scopeRes)) {
+      return false;
+    }
+
+    Variables locals;
+    if (!client.GetLocals(frame, &locals)) {
+      return false;
+    }
+
+    for (int i = 0;; i++) {
+      auto lane_var = client.GetLane(locals, i);
+      if (!lane_var) {
+        break;
+      }
+      if (auto var = lane_var->Find(kGlobalInvocationId)) {
+        std::tuple<uint32_t, uint32_t, uint32_t> got;
+        if (var->Get(&got)) {
+          uint32_t x;
+          uint32_t y;
+          uint32_t z;
+          std::tie(x, y, z) = got;
+          if (x == id.x && y == id.y && z == id.z) {
+            *lane = i;
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
+  }
+
+  // FindVertexIndex looks for the requested vertex shader's vertex index in the
+  // stack frames' locals, returning true if found, and assigns the index of the
+  // SIMD lane it was found in to |lane|.
+  // TODO(bclayton): This value should probably be in the globals, not locals!
+  bool FindVertexIndex(dap::integer threadId, uint32_t index, int* lane) {
+    Client client(session_, [this](const std::string& err) { OnError(err); });
+
+    dap::StackFrame frame;
+    if (!client.TopStackFrame(threadId, &frame)) {
+      return false;
+    }
+
+    dap::ScopesRequest scopeReq;
+    dap::ScopesResponse scopeRes;
+    scopeReq.frameId = frame.id;
+    if (!client.Send(scopeReq, &scopeRes)) {
+      return false;
+    }
+
+    Variables locals;
+    if (!client.GetLocals(frame, &locals)) {
+      return false;
+    }
+
+    for (int i = 0;; i++) {
+      auto lane_var = client.GetLane(locals, i);
+      if (!lane_var) {
+        break;
+      }
+      if (auto var = lane_var->Find(kVertexIndex)) {
+        uint32_t got;
+        if (var->Get(&got)) {
+          if (got == index) {
+            *lane = i;
+            return true;
+          }
+        }
+      }
+    }
+
+    return false;
+  }
+
+  void OnError(const std::string& error) {
+    DEBUGGER_LOG("ERROR: %s", error.c_str());
+    error_ += error;
+  }
+
+  using PendingThreadsMap = std::unordered_map<InvocationKey,
+                                               debug::Script::OnThread,
+                                               InvocationKey::Hash>;
+  using ThreadVector = std::vector<std::unique_ptr<Thread>>;
+  std::shared_ptr<dap::Session> session_;
+  std::mutex threads_mutex_;
+  PendingThreadsMap pendingThreads_;  // guarded by threads_mutex_
+  ThreadVector runningThreads_;       // guarded by threads_mutex_
+  std::mutex error_mutex_;
+  Result error_;  // guarded by error_mutex_
+};
+
+std::pair<Engine::Debugger*, Result> EngineVulkan::GetDebugger() {
+  if (!debugger_) {
+    auto debugger = new VkDebugger();
```

Comment:
any reason to not `debugger_ = MakeUnique<Debugger>()` ? This currently leaks if you early return on line 954.

Reply:
Yeah, this downcasts and I'd need to upcast again to call the `Connect()` method.
Reshuffled things so that the leak shouldn't happen. Good spot!